### DISCUSSION TITLE
How to make single executable file from pre cython build binary files ?

### DISCUSSION THREAD

**Discussion by gaamaaresosa** [pyinstaller/pyinstaller#7638] [PyInstaller]
I know PyInstaller in the only easy and 100% working executable maker for python project.
For security reason I need to Cythonise all my MyMainApp.py file and other shared python modules into *.so files by cython.
MyMainApp.so (Shared Object binary from by cython)
OtherSharedObject1.so (Shared Object binary from by cython)
OtherSharedObject2.so (Shared Object binary from by cython)
OtherSharedObject3.so (Shared Object binary from by cython)

If I call below command line I get a single binary file EntryMain.
So far everything is ok.

`
pyinstaller -c -F --clean EntryMain.py
`

EntryMain.py
```
import MyMainApp # this is where your def main function is located, the method, that actually starts your application.

if __name__ == "__main__":
   MyMainApp .main() # asuming that your method is called main.
```

But I don't want to make my MyMainApp.py into shared object (MyMainApp.so)
I need to make an executable binary from cython with embed command.
So I made the MyMainApp executable binary successfully from cython.

Now how to make a single executable file from PyInstaller by using below files ?
MyMainApp (Executable binary from by cython)
OtherSharedObject1.so (Shared Object binary from by cython)
OtherSharedObject2.so (Shared Object binary from by cython)
OtherSharedObject3.so (Shared Object binary from by cython)

My guess:
If I create a required_imports.py as below and import this in my EntryMain.py
And how to exec MyMainApp executable file from if __name__ == "__main__": ?

required_imports.py
```
# Import this file into your python entrypoint to get PyInstaller to recognize the imports required by your program.
import OtherSharedObject1
import OtherSharedObject2
import OtherSharedObject3
```

Can some one help me with step by step explanation ?

My system:
Raspberry Pi-4
32bit Linux Buster

**Comment by bwoodsend** [comment]
We are not providing support for this. Running Cythonized code through PyInstaller is very much an advanced route where getting it built and running on your own machine (i.e. the step you're on) is the first and easiest step.

I suggest you try pyarmor instead (although be aware that it's just an extreme version of our deprecated --key option).

**Reply by gaamaaresosa** [reply]
@bwoodsend 
Thank you so much for your reply.
Forgive me if I still talk as less knowledge on python and PyInstaller.

> We are not providing support for this. Running Cythonized code through PyInstaller 

You may forget about cython or cythonized keywords.
I am stuck with how to call an executable from Python code.
As I am not a python expert.

Eg: EntryMain.py
```
if __name__ == "__main__":
      shell MyMainApp #(In Windows we call like this)
```
If I add my files via PyInstaller Hidden Import or some other methods
And pass my above EntryMain.py to PyInstaller, then PyInstaller should execute the comand right ?
Then rest will be taken care by my executable because all the files are available on the PyInstaller VM path right ?

> I suggest you try pyarmor instead (although be aware that it's just an extreme version of our deprecated --key option).

Seems pyarmor also can be reversed easily.

**Reply by bwoodsend** [reply]
What executable are you trying to call from Python code? You appear to be saying that you're running your main application as a subprocess? That is definitely not a sensible thing to do.

**Reply by gaamaaresosa** [reply]
> What executable are you trying to call from Python code? You appear to be saying that you're running your main application as a subprocess? That is definitely not a sensible thing to do.

I agree...

I don't know how to proceed as I am stuck on my issue for many weeks.

Example:
I have 1 executable binary "MainApp" made from cython. (Its runs fine with **sudo ./MainApp**)
Then I have 3 shared objects (*.so) made from cython again.
These shared objects already imported on the MainApp.
If I put all these files on a folder and execute MainApp then all works fine.

Now I need a single executable from these all 4 files.
Since PyInstaller is a good bundler, I some how try using it.
You can suggest me any other bundler also.
Which should extract these 4 file on memory and run the project well.
Here python or other dependencies are not matter, because already everyting is present on the target linux box.

**Reply by rokm** [reply]
Surely you can do a google search for "linux self-extracting archive" and find a better solution than trying to shoehorn this into PyInstaller's onefile build?

**Reply by gaamaaresosa** [reply]
> Surely you can do a google search for "linux self-extracting archive" and find a better solution than trying to shoehorn this into PyInstaller's onefile build?

Thanks for the good key word to search.
I will try.
